peregrine/resources/submission/graphql/node.py

```diff
@@ -788,3 +786,97 @@ def get_fields():
 
 
 NodeField = graphene.List(Node, args=get_node_interface_args())
+
+
+# ======================================================================
+# DataNode
+
+
+class DataNode(graphene.Interface):
+    id = graphene.ID()
+    shared_fields = None
+
+
+# return a dictionary containing the fields shared by all data nodes
+def get_shared_fields_dict():
+    if not DataNode.shared_fields:
+        shd_fields = None
+        for node in dictionary.schema:
```

Comment:
This section could be condensed a little bit using comprehensions; something like this
```python
# fields lists the set of node fields, for every data node in the dictionary schema
# (nodes ending with '_file')
fields = [
    set(schema['properties'].keys())
    for schema in dictionary.schema.values()
    if schema['category'].endswith('_file')
]
# shared_fields takes the intersection of all the data node field sets
shared_fields = set.intersection(*fields)

shared_fields_dict = {field: graphene.String() for field in shared_fields}
if 'file_size' in shared_fields:
    shared_fields_dict['file_size'] = graphene.Int()
```